### PR TITLE
[ML] Ensure immutability of MlMetadata

### LINKED ISSUE #31916: [ML] [CI] DetectionRulesIT.testCondition failure
This test fails with the error:

```
Expected: "low"
     but: was "high"
	at __randomizedtesting.SeedInfo.seed([C579EF26625EB1F4:9CC01CBE66771FFC]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.xpack.ml.integration.DetectionRulesIT.testCondition(DetectionRulesIT.java:120)
```

It appears the updated rules are not being picked up. 

Reproduces fairly frequently for me ~ every other time:

```
./gradlew :x-pack:qa:ml-native-multi-node-tests:integTestRunner \
  -Dtests.seed=C579EF26625EB1F4 \
  -Dtests.class=org.elasticsearch.xpack.ml.integration.DetectionRulesIT \
  -Dtests.method="testCondition" \
  -Dtests.security.manager=true \
  -Dtests.locale=pt \
  -Dtests.timezone=Etc/GMT-7
```

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
The test failure in #31916 revealed that updating
rules on a job was modifying the detectors list
in-place. That meant the old cluster state and the
updated cluster state had no difference and thus the
change was not propagated to non-master nodes.

This commit fixes that and also reviews all of ML
metadata in order to ensure immutability.

Closes #31916

### CODE REVIEW

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/job/config/JobUpdateTests.java [hunk 1]
```diff
@@ -172,12 +173,10 @@ public void testMergeWithJob() {
...
         for (JobUpdate.DetectorUpdate detectorUpdate : update.getDetectorUpdates()) {
-            assertNotNull(updatedJob.getAnalysisConfig().getDetectors().get(detectorUpdate.getDetectorIndex()).getDetectorDescription());
-            assertEquals(detectorUpdate.getDescription(),
-                    updatedJob.getAnalysisConfig().getDetectors().get(detectorUpdate.getDetectorIndex()).getDetectorDescription());
-            assertNotNull(updatedJob.getAnalysisConfig().getDetectors().get(detectorUpdate.getDetectorIndex()).getDetectorDescription());
-            assertEquals(detectorUpdate.getRules(),
-                    updatedJob.getAnalysisConfig().getDetectors().get(detectorUpdate.getDetectorIndex()).getRules());
+            Detector updatedDetector = updatedJob.getAnalysisConfig().getDetectors().get(detectorUpdate.getDetectorIndex());
+            assertNotNull(updatedDetector);
+            assertEquals(detectorUpdate.getDescription(), updatedDetector.getDetectorDescription());
```
Comment (davidkyle) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/job/config/JobUpdateTests.java]: Can you also assert that the original job has not been changed

**Review by davidkyle — COMMENTED** [review]
Looks good. I would like to add more tests if possible perhaps in AnalysisConfigTests::mutateInstance as if detector rules where changed there we might have caught the bug

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java [hunk 2]
```diff
@@ -156,14 +156,14 @@ private DatafeedConfig(String id, String jobId, TimeValue queryDelay, TimeValue
...
+        this.indices = indices == null ? null : Collections.unmodifiableList(indices);
+        this.types = types == null ? null : Collections.unmodifiableList(types);
         this.query = query;
         this.aggregations = aggregations;
-        this.scriptFields = scriptFields;
+        this.scriptFields = scriptFields == null ? null : Collections.unmodifiableList(scriptFields);
         this.scrollSize = scrollSize;
         this.chunkingConfig = chunkingConfig;
-        this.headers = Objects.requireNonNull(headers);
+        this.headers = headers == null ? null : Collections.unmodifiableMap(headers);
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java]: `headers` should still not be allowed to be `null`.

**Comment by dimitris-athanasiou** [comment]
I have added tests that attempt to ensure immutability of jobs/datafeeds, at least with regard to updatable properties. Could you have a look at those and let me know if you think they're worth it?

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**